### ISSUE TITLE
StackOverflowException in core interpreter with stream shapes involving nested bind/uncons

### ISSUE THREAD

**Issue by mpilquist** [functional-streams-for-scala/fs2#1107]
Reproduction case:

```scala
package example

import fs2._
import cats.effect.IO
import scala.concurrent.ExecutionContext.Implicits.global

object ObserveSoe extends App {
  val stream = Stream
    .range(0, 100000)
    .unchunk                    // Create lots of segments
    .covary[IO]
    .prefetch                   // Push segments through a queue
    .flatMap(_ => Stream.empty) // Generate a bind node that emits no elements
    .drain                      // Add a combinator that pulls a segment
  stream.compile.drain.unsafeRunSync
}
```

**Comment by mpilquist** [comment]
Confirmed with some trace logging that the SOE is here: https://github.com/functional-streams-for-scala/fs2/blob/ed15bbc569ff81b01b276ec51ce1f00cfc6691aa/core/shared/src/main/scala/fs2/internal/Algebra.scala#L267-L270

**Comment by SystemFw** [comment]
Thanks for minimising this, it was driving me insane :)

**Comment by pchlupacek** [comment]
ok I did a bit of investigation there, and this is what I found : 

It seems to related only to queue, and particularly dequeue, which is defined as : 
```scala 
Stream
      .bracket(cancellableDequeue1)(d => Stream.eval(d._1), d => d._2)
      .repeat
```
now the issue seems to be related to bracket which is defined as 
```scala 
fromFreeC(Algebra.acquire[F, O, R](r, release).flatMap {
      case (r, token) =>
        use(r).onComplete { fromFreeC(Algebra.release(token)) }.get
    })

```
and particularly the `onComplete` which is defined as: 

```scala
(self.handleErrorWith(e => s2 ++ Stream.raiseError(e))) ++ s2
```
note tha since 0.10 the hadleError is defined as (note the Scope): 

```scala
 fromFreeC(Algebra.scope(self.get[F, O2]).handleErrorWith { e =>
        h(e).get[F, O2]
      })
```
as such I think it also uncovers the problem with queue, which when it takes one element only, 
it will in fact always open/close scope as result.

**Comment by pchlupacek** [comment]
Just to clear things up, as it may not be obvious, every `bracket` now opens and closes scope, making the Acquire/Release sort of redundant from the scope now.

**Comment by mpilquist** [comment]
@pchlupacek I probably don't fully understand the issue yet, but shouldn't the interpreter support the current queue.dequeue scenario, even if it could/should be optimized? And shouldn't we be able to reproduce without queue by replacing the prefetch with something that creates a scope in a similar way?

**Comment by pchlupacek** [comment]
of course, I am not saying otherwise. Sorry not be clear. Is just that I found under changes that were made the `bracket` and consequently `dequeue` do have much higher cost than before thanks to change introduced to `handleErrorWith`. 

I am still not convinced on the bug nature, this was just to put it more in perspective. So far my theory on whats happening is that we accumulate the tail of the `Uncons` with each `bracket`, that means always when we start interpreting Uncons-ed stream and you leave to interpret Scope related Algebra. That I think accumulates functions that results in SoE and in fact are causing the memory leak.

**Comment by mpilquist** [comment]
@pchlupacek Here's a more minimal reproduction based on your findings, showing the interplay between scopes, streams that don't emit, and uncons nodes:

```scala
package example

import fs2._
import cats.effect.IO

object ObserveSoe extends App {
  Stream(0)
    .covary[IO]
    .scope                      // Create a source that opens/closes a new scope for every element emitted
    .repeat
    .take(10000)
    .flatMap(_ => Stream.empty) // Never emit an element downstream
    .drain                      // Use a combinator that calls Stream#pull.uncons
    .compile
    .drain
    .unsafeRunSync
}
```

**Comment by pchlupacek** [comment]
@mpilquist excellent. thanks

**Comment by mpilquist** [comment]
Fixed by #1110
